### ISSUE TITLE
File changes in PR

### ISSUE THREAD

**Issue by ankitbko** [MicrosoftDocs/vsts-rest-api-specs#68]
Using API, how do I get files changed in a PR. The [PR API](https://docs.microsoft.com/en-us/rest/api/vsts/git/pull%20requests/get%20pull%20request?view=vsts-rest-4.1#gitpullrequest) should return `GitPullRequest` which should contain `GitCommitRef[]`. As per the document [GitCommitRef](https://docs.microsoft.com/en-us/rest/api/vsts/git/pull%20requests/get%20pull%20request?view=vsts-rest-4.1#gitcommitref) should contain `GitChange[] ` where each `GitChange` will show what changed.

However none of this comes when sending a Get request to PR endpoint. Even with `includeCommits=true` no actual changes are shown.

I also tried to use the [diff](https://docs.microsoft.com/en-us/rest/api/vsts/git/diffs/get?view=vsts-rest-4.1) endpoint to get changes between branches but to no avail. This is the response I get from diff endpoint which is utterly useless.
```
{"allChangesIncluded":true,"changeCounts":{},"changes":[],"commonCommit":"c03a8711ea215a80f9283e10d104eebccd17bd43","baseCommit":"76ea7f4ac6d008a7947d810f296ae360317e9e33","targetCommit":"152314ddc8bf05d12d871f381cb500489ebbc371","aheadCount":1,"behindCount":83}
```

Simply put - how do I get list of files changed and changes between them in a PR?

**Comment by mmitrik** [comment]
It's not obvious from the docs, but you need to use the [PR Iterations API](https://docs.microsoft.com/en-us/rest/api/vsts/git/pull%20request%20iterations?view=vsts-rest-4.1).  Use this to list the iterations in the PR, then you want to get the number of the latest iteration.  Then, you can use [PR Iteration Changes](https://docs.microsoft.com/en-us/rest/api/vsts/git/pull%20request%20iteration%20changes/get?view=vsts-rest-4.1) to get the list of changes for an iteration, or between two iterations.  If you want all changes, use the latest iteration and set the `compareTo=0`.

Once you have the list of changes, you need to use the [Items API](https://docs.microsoft.com/en-us/rest/api/vsts/git/items?view=vsts-rest-4.1) to get the actual contents.

Hopefully that helps until we can get some better samples posted.  

cc @steved0x

**Comment by ankitbko** [comment]
@mmitrik Thanks. I can get list of files changed in the PR using iteration changes. But I am still not able to get the diff between files. The Items api just returns be the file for a given commit. Is there a way I can see *what* changed in a file (lines added/removed/modified)?

**Comment by ankitbko** [comment]
So I was working on trying to create an extension for vscode that integrates with AZDO and wanted to get the diff of the files before I landed on this issue. Don't even remember when I raised it 😸. Well anyways, is there an update for this? @mmitrik You would have moved on to other roles so can you recommend someone I can connect with directly?

cc: @tedchamb since you are the top contributor.

**Comment by mmitrik** [comment]
Not positive of the best person for this.  @aaronhallberg might be able to advise.

**Comment by tedchamb** [comment]
@goflores, who would be the best person to help with this REST API question?

**Comment by alexanderkovshov** [comment]
@ankitbko Try to swap target and base versions, that will do the trick. But it won't show you the diff, only path of changed and change type, you can check samples [here](https://docs.microsoft.com/en-us/rest/api/azure/devops/git/diffs/get?view=azure-devops-rest-6.0) 

Use this to get diff:
Step1: Use the Rest API to get commit id from PR.

GET https://dev.azure.com/{organization}/{project}/_apis/git/repositories/{repositoryId}/diffs/commits?api-version=6.0

Step2: You could use the Rest API to get the commit by commit id.

GET https://dev.azure.com/{organization}/{project}/_apis/git/repositories/{repositoryId}/commits/{commitId}?changeCount={changeCount}&api-version=5.0
In the Rest API Result, you need to record the value of “parentsid”, “path”.

Step3: You could use the following API to get the diff content.

Post https://dev.azure.com/{organization}/{project}/_api/_versioncontrol/fileDiff?__v=5&diffParameters={value}&repositoryId={repositoryid}
The {value} is Json type.

Here is an example:

{"originalPath":"filepath","originalVersion":"Parentsid","modifiedPath":"filepath","modifiedVersion":"commitid","partialDiff":true,"includeCharDiffs":true}
You could add the value to the API URL.

**Comment by ankitbko** [comment]
Thank you. I didnt see fileDiff api in the documentation though.

**Comment by AshwaniT** [comment]
_api/_versioncontrol/fileDif is gioving 404 for my instance.
